File: cpp/test/data/test_eop.cc
#include <lupnt/conversions/frame_conversions.h>
#include <lupnt/conversions/time_conversions.h>
#include <lupnt/core/constants.h>
#include <lupnt/core/file.h>
#include <lupnt/interfaces/eop.h>

#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>
#include <cmath>
#include <filesystem>
#include <string>
#include <vector>

#include "../utils.cc"

using namespace lupnt;
using namespace Catch::Matchers;

const double epsilon = 1e-6;

TEST_CASE("data.eop") {
  EopFileData* file_data = GetEopFileData();
  REQUIRE(file_data != nullptr);
  REQUIRE(file_data->mjds_utc.size() > 10);
  REQUIRE(file_data->mjds_utc(1) > file_data->mjds_utc(0));

  Real mjd = file_data->mjds_utc(file_data->mjds_utc.size() / 2);
  EopData data = GetEopData(mjd);
  REQUIRE(std::isfinite(data.x_pole.val()));
  REQUIRE(std::isfinite(data.y_pole.val()));
  REQUIRE(std::isfinite(data.ut1_utc.val()));
  REQUIRE(std::isfinite(data.lod.val()));
  REQUIRE_THAT(GetUt1UtcDifference(mjd).val(), WithinAbs(data.ut1_utc.val(), epsilon));
}

// ----------------------------------------------------------------------------
// Coverage reporting and the clamp-outside-the-table behavior. The clamp is the
// dominant Earth-orientation error for simulations run at epochs past the end of
// the loaded file, so it is pinned here rather than left implicit.
// ----------------------------------------------------------------------------
TEST_CASE("data.eop_coverage") {
  EopFileData* file_data = GetEopFileData();
  EopCoverage cov = GetEopCoverage();

  REQUIRE_THAT(cov.mjd_first, WithinAbs(file_data->mjds_utc(0), epsilon));
  REQUIRE_THAT(cov.mjd_last,
               WithinAbs(file_data->mjds_utc(file_data->mjds_utc.size() - 1), epsilon));
  REQUIRE(cov.mjd_last > cov.mjd_first);

  // Past the end of the table: the last row is held constant, not extrapolated.
  int last = file_data->mjds_utc.size() - 1;
  EopData beyond = GetEopData(Real(cov.mjd_last + 365.0));
  REQUIRE_THAT(beyond.x_pole.val(), WithinAbs(file_data->x(last) * RAD_ARCSEC, epsilon));
  REQUIRE_THAT(beyond.y_pole.val(), WithinAbs(file_data->y(last) * RAD_ARCSEC, epsilon));
  REQUIRE_THAT(beyond.ut1_utc.val(), WithinAbs(file_data->ut1_utc(last), epsilon));

  // Before the start: likewise held at the first row.
  EopData before = GetEopData(Real(cov.mjd_first - 365.0));
  REQUIRE_THAT(before.x_pole.val(), WithinAbs(file_data->x(0) * RAD_ARCSEC, epsilon));
  REQUIRE_THAT(before.ut1_utc.val(), WithinAbs(file_data->ut1_utc(0), epsilon));

  // The sigma_* fields are the C04 formal errors, not CIP offsets or TAI-UTC.
  REQUIRE_THAT(before.sigma_x_pole.val(), WithinAbs(file_data->xErr(0) * RAD_ARCSEC, epsilon));
  REQUIRE_THAT(before.sigma_y_pole.val(), WithinAbs(file_data->yErr(0) * RAD_ARCSEC, epsilon));
  REQUIRE_THAT(before.sigma_ut1_utc.val(), WithinAbs(file_data->ut1_utc_err(0), epsilon));
}

// ----------------------------------------------------------------------------
// The IERS finals ("Bulletin A") reader, exercised offline against the finals.all snapshot
// bundled with the repo's orekit data. Restores the bundled C04 table afterwards so the rest of
// the suite (calibrated against that vintage) is unaffected.
// ----------------------------------------------------------------------------
TEST_CASE("data.eop_finals") {
  std::filesystem::path finals = std::filesystem::path(GetDataPath()) / ".." / "orekit-data-main"
                                 / "Earth-Orientation-Parameters" / "IAU-1980" / "finals.all";
  if (!std::filesystem::exists(finals)) {
    WARN("orekit finals.all snapshot not present; skipping finals reader test");
    return;
  }

  LoadEopFinalsFileData(finals, true);
  EopFileData* fd = GetEopFileData();
  EopCoverage cov = GetEopCoverage();

  REQUIRE(fd->source == EopSource::Finals);
  REQUIRE(fd->mjds_utc.size() > 10000);
  REQUIRE(fd->is_prediction.size() == fd->mjds_utc.size());
  // The series begins 1973-01-02 (MJD 41684) and is strictly daily.
  REQUIRE_THAT(cov.mjd_first, WithinAbs(41684.0, epsilon));
  REQUIRE_THAT(fd->mjds_utc(1) - fd->mjds_utc(0), WithinAbs(1.0, epsilon));
  REQUIRE(fd->years(0) == 1973);
  REQUIRE(fd->months(0) == 1);
  REQUIRE(fd->days(0) == 2);

  // First row, checked field by field against the raw file (units: x/y in arcsec, UT1-UTC in s,
  // LOD converted from ms, dpsi/deps converted from mas).
  // 73 1 2 41684.00 I  0.120733 0.009786  0.136966 0.015902  I 0.8084178 0.0002710  0.0000 0.1916
  REQUIRE_THAT(fd->x(0), WithinAbs(0.120733, epsilon));
  REQUIRE_THAT(fd->y(0), WithinAbs(0.136966, epsilon));
  REQUIRE_THAT(fd->ut1_utc(0), WithinAbs(0.8084178, epsilon));
  REQUIRE_THAT(fd->xErr(0), WithinAbs(0.009786, epsilon));
  REQUIRE_THAT(fd->yErr(0), WithinAbs(0.015902, epsilon));
  REQUIRE_THAT(fd->ut1_utc_err(0), WithinAbs(0.0002710, epsilon));
  REQUIRE_THAT(fd->lod_err(0), WithinAbs(0.1916e-3, epsilon));
  REQUIRE(fd->is_prediction(0) == 0);

  // The table must contain both measured and predicted rows, in that order, and the predictions
  // must extend past the measured span -- that reach is the whole point of using finals.
  REQUIRE(cov.mjd_last_measured > cov.mjd_first);
  REQUIRE(cov.mjd_last > cov.mjd_last_measured);
  REQUIRE(fd->is_prediction(0) == 0);
  REQUIRE(fd->is_prediction(fd->is_prediction.size() - 1) == 1);
  for (Eigen::Index i = 1; i < fd->mjds_utc.size(); ++i) {
    REQUIRE(fd->mjds_utc(i) > fd->mjds_utc(i - 1));                         // strictly increasing
    if (fd->is_prediction(i - 1) == 1) REQUIRE(fd->is_prediction(i) == 1);  // no measured-after-P
    REQUIRE((fd->mjds_utc(i) <= cov.mjd_last_measured) == (fd->is_prediction(i) == 0));
  }

  // Values must stay physical across the whole table, predictions included: polar motion within
  // a second of arc, |UT1-UTC| under a second (leap seconds keep it there), LOD sub-millisecond.
  for (Eigen::Index i = 0; i < fd->mjds_utc.size(); ++i) {
    REQUIRE(std::abs(fd->x(i)) < 1.0);
    REQUIRE(std::abs(fd->y(i)) < 1.0);
    REQUIRE(std::abs(fd->ut1_utc(i)) < 1.0);
    REQUIRE(std::abs(fd->lod(i)) < 0.01);
  }

  // Interpolating in the predicted span works and returns finite values.
  EopData predicted = GetEopData(Real(0.5 * (cov.mjd_last_measured + cov.mjd_last)));
  REQUIRE(std::isfinite(predicted.x_pole.val()));
  REQUIRE(std::isfinite(predicted.ut1_utc.val()));
  REQUIRE(std::isfinite(predicted.lod.val()));

  // Finals must agree with the independent C04 series where both are final. Compare at an epoch
  // well inside both (MJD 59000, 2020-05-31); tolerances are a few mas / tens of microseconds,
  // which is the genuine C04-vs-Bulletin-A solution difference.
  Real mjd_common = 59000.0;
  EopData f = GetEopData(mjd_common);
  LoadEopFileData(GetFilePath(EOP_FILENAME), true);
  EopData c = GetEopData(mjd_common);
  REQUIRE_THAT((f.x_pole - c.x_pole).val() / RAD_ARCSEC, WithinAbs(0.0, 5e-3));  // < 5 mas
  REQUIRE_THAT((f.y_pole - c.y_pole).val() / RAD_ARCSEC, WithinAbs(0.0, 5e-3));
  REQUIRE_THAT((f.ut1_utc - c.ut1_utc).val(), WithinAbs(0.0, 1e-4));  // < 0.1 ms
  REQUIRE_THAT((f.lod - c.lod).val(), WithinAbs(0.0, 1e-4));

  // ...and the restored table is C04 again.
  REQUIRE(GetEopFileData()->source == EopSource::C04);
}

// ----------------------------------------------------------------------------
// SetEopSource: the selection must be independent of call order. This is the defect it exists to
// fix -- a bare Load*(path, force=false) issued after anything has touched Earth orientation is
// dropped, so a run can believe it selected one product while carrying another.
// ----------------------------------------------------------------------------
TEST_CASE("data.eop_source_selection") {
  std::filesystem::path finals = std::filesystem::path(GetDataPath()) / ".." / "orekit-data-main"
                                 / "Earth-Orientation-Parameters" / "IAU-1980" / "finals.all";
  if (!std::filesystem::exists(finals)) {
    WARN("orekit finals.all snapshot not present; skipping source selection test");
    return;
  }

  // Default is C04.
  LoadEopFileData(GetFilePath(EOP_FILENAME), true);
  REQUIRE(GetEopSource() == EopSource::C04);
  REQUIRE(GetEopCoverage().source == EopSource::C04);

  // Selecting with a table already loaded switches immediately rather than deferring (which
  // would amount to ignoring the call).
  SetEopSource(EopSource::Finals, finals);
  REQUIRE(GetEopSource() == EopSource::Finals);
  REQUIRE(GetEopCoverage().source == EopSource::Finals);
  REQUIRE(GetEopFileData()->source == EopSource::Finals);

  // Selecting the source already in force is a no-op, not a reload.
  double mjd_last = GetEopCoverage().mjd_last;
  SetEopSource(EopSource::Finals, finals);
  REQUIRE(GetEopCoverage().mjd_last == mjd_last);

  // ...and back.
  SetEopSource(EopSource::C04);
  REQUIRE(GetEopCoverage().source == EopSource::C04);

  // Selecting a source whose file is absent fails at the point of the call, not later inside a
  // frame conversion.
  REQUIRE_THROWS(SetEopSource(EopSource::Finals, finals.parent_path() / "no_such_finals_file"));
  // ...and a failed selection leaves the previous choice intact.
  REQUIRE(GetEopSource() == EopSource::C04);
  REQUIRE(GetEopCoverage().source == EopSource::C04);

  // A non-forcing load with a table already present is dropped -- the behavior SetEopSource
  // exists to route around. Pinned here so it cannot regress into a silent switch.
  LoadEopFinalsFileData(finals, false);
  REQUIRE(GetEopCoverage().source == EopSource::C04);

  LoadEopFileData(GetFilePath(EOP_FILENAME), true);
}

// ----------------------------------------------------------------------------
// Loading the latest EOP series from the IERS data center. This test does not
// require network access: if the download fails (offline environment), it
// just verifies that the bundled EOP data remains usable.
// ----------------------------------------------------------------------------
TEST_CASE("data.eop_latest_iers") {
  double mjd_last_before = GetEopFileData()->mjds_utc(GetEopFileData()->mjds_utc.size() - 1);

  bool downloaded = LoadLatestEopFromIers(true);

  EopFileData* file_data = GetEopFileData();
  REQUIRE(file_data != nullptr);
  REQUIRE(file_data->mjds_utc.size() > 10);

  if (downloaded) {
    double mjd_last_after = file_data->mjds_utc(file_data->mjds_utc.size() - 1);
    REQUIRE(mjd_last_after > mjd_last_before);

    // 2025-01-01 in MJD, well within the latest IERS series.
    Real mjd_recent = 60676.0;
    EopData data = GetEopData(mjd_recent);
    REQUIRE(std::isfinite(data.x_pole.val()));
    REQUIRE(std::isfinite(data.y_pole.val()));
    REQUIRE(std::isfinite(data.ut1_utc.val()));

    // Restore the bundled EOP data so other tests (e.g. ITRF cross-validation,
    // which is calibrated against the bundled EOP vintage) are unaffected.
    LoadEopFileData(GetFilePath(EOP_FILENAME), true);
  } else {
    Real mjd = file_data->mjds_utc(file_data->mjds_utc.size() / 2);
    EopData data = GetEopData(mjd);
    REQUIRE(std::isfinite(data.x_pole.val()));
    REQUIRE(std::isfinite(data.ut1_utc.val()));
  }
}

// ----------------------------------------------------------------------------
// EOP perturbation injector (Step 2a). Pins the small-angle convention by central difference
// rather than asserting it -- the signs of eps are exactly the kind of thing that is easy to get
// backwards and impossible to notice downstream.
//
//   delta r_gcrf = R_itrf->gcrf * (eps x r_itrf),
//   eps = (-dy_pole, -dx_pole, OMEGA_ERA * dut1)   [rad, in ITRF]
//
// Note the UT1 step: LuPNT epochs are seconds since J2000 (~1.9e9), so a 1e-6 s step is at
// double-precision resolution and the difference is pure roundoff. 1e-2 s is needed.
// ----------------------------------------------------------------------------
TEST_CASE("data.eop_perturbation") {
  LoadEopFileData(GetFilePath(EOP_FILENAME), true);

  // No perturbation by default, and the fast-path flag is off for a C04 table.
  ClearEopPerturbation();
  REQUIRE(EopHasCelestialPoleOffsets() == false);
  EopPerturbation none = GetEopPerturbation(Real(59000.0));
  REQUIRE(none.dut1.val() == 0.0);

  Real mjd = 59000.0;
  EopData nominal = GetEopData(mjd);

  // A constant perturbation must land additively on every field it names.
  EopPerturbation p;
  p.dx_pole = 1e-8;
  p.dy_pole = -2e-8;
  p.dut1 = 3e-3;
  p.dlod = 4e-6;
  SetEopPerturbation(p);
  REQUIRE(EopHasCelestialPoleOffsets() == true);
  EopData perturbed = GetEopData(mjd);
  REQUIRE_THAT((perturbed.x_pole - nominal.x_pole).val(), WithinAbs(1e-8, 1e-14));
  REQUIRE_THAT((perturbed.y_pole - nominal.y_pole).val(), WithinAbs(-2e-8, 1e-14));
  REQUIRE_THAT((perturbed.ut1_utc - nominal.ut1_utc).val(), WithinAbs(3e-3, 1e-9));
  REQUIRE_THAT((perturbed.lod - nominal.lod).val(), WithinAbs(4e-6, 1e-12));

  // A time-varying perturbation is evaluated at the epoch requested, not at a fixed one.
  ClearEopPerturbation();
  EopData nominal_10 = GetEopData(Real(59010.0));
  SetEopPerturbationFunction([](Real m) {
    EopPerturbation q;
    q.dut1 = (m - 59000.0) * 1e-3;
    return q;
  });
  REQUIRE_THAT((GetEopData(Real(59010.0)).ut1_utc - nominal_10.ut1_utc).val(),
               WithinAbs(1e-2, 1e-9));
  REQUIRE_THAT((GetEopData(Real(59000.0)).ut1_utc - nominal.ut1_utc).val(), WithinAbs(0.0, 1e-12));
  ClearEopPerturbation();
  REQUIRE(EopHasCelestialPoleOffsets() == false);
  REQUIRE(GetEopPerturbation(Real(59010.0)).dut1.val() == 0.0);

  // ---- the small-angle convention, by central difference on GCRF<->ITRF ----
  Real t_tdb = ConvertTime(MjdToTime(mjd), Time::UTC, Time::TDB);
  Vec3 r_itrf(4197160.8, 815845.4, 4716876.3);

  auto rotated = [&](const EopPerturbation& q) {
    SetEopPerturbation(q);
    Vec3 out = ItrfToGcrf(t_tdb, r_itrf);
    ClearEopPerturbation();
    return out;
  };
  Vec3 r_nominal = rotated(EopPerturbation{});
  // No public accessor for the composite rotation, so build it from the public conversion:
  // ITRF->GCRF is a pure rotation (both geocentric), so its columns are the images of the axes.
  Mat3 R;
  R.col(0) = ItrfToGcrf(t_tdb, Vec3(1, 0, 0));
  R.col(1) = ItrfToGcrf(t_tdb, Vec3(0, 1, 0));
  R.col(2) = ItrfToGcrf(t_tdb, Vec3(0, 0, 1));

  struct Case {
    const char* name;
    double step;
    Vec3 axis;  // eps per unit of the parameter, in ITRF
    EopPerturbation (*make)(double);
  };
  const double OMEGA_ERA = TWO_PI * 1.00273781191135448 / SECS_DAY;
  std::vector<Case> cases{
      {"dx_pole", 1e-8, Vec3(0, -1, 0),
       [](double h) {
         EopPerturbation q;
         q.dx_pole = h;
         return q;
       }},
      {"dy_pole", 1e-8, Vec3(-1, 0, 0),
       [](double h) {
         EopPerturbation q;
         q.dy_pole = h;
         return q;
       }},
      {"dut1", 1e-2, Vec3(0, 0, 1),
       [](double h) {
         EopPerturbation q;
         q.dut1 = h;
         return q;
       }},
  };
  for (const Case& c : cases) {
    Vec3 num = (rotated(c.make(c.step)) - rotated(c.make(-c.step))) / (2.0 * c.step);
    Vec3 eps = c.axis;
    if (std::string(c.name) == "dut1") eps *= OMEGA_ERA;
    Vec3 ana = R * eps.cross(r_itrf);
    double rel = (num - ana).norm().val() / ana.norm().val();
    INFO("parameter " << c.name << " rel.err " << rel);
    REQUIRE(rel < 1e-4);
  }

  // The unperturbed rotation must be exactly restored.
  REQUIRE_THAT((rotated(EopPerturbation{}) - r_nominal).norm().val(), WithinAbs(0.0, 1e-9));
  ClearEopPerturbation();
}

File: python/test/test_eop.py
"""Unit tests for the `pylupnt` Earth-orientation bindings (`get_eop_coverage`,
`get_eop_data`, `get_eop_file_data`, `load_eop_file_data`,
`load_eop_finals_file_data`).

These exercise the two EOP sources through their Python bindings: the bundled
IERS C04 final series and the IERS finals ("Bulletin A") reader, which is the
only one carrying predictions. Network-dependent loaders
(`load_latest_eop_*_from_iers`) are deliberately not exercised here.

Every test restores the bundled C04 table on the way out, since the table is
global process state and the rest of the suite is calibrated against it.
"""

import os

import numpy as np
import pylupnt as pnt
import pytest

C04_FILENAME = "eopc04_08.62-now"

# The repo's orekit data snapshot ships an IAU-1980 finals.all, used here so the
# finals reader can be tested without network access.
FINALS_PATH = os.path.join(
    pnt.get_data_path(),
    "..",
    "orekit-data-main",
    "Earth-Orientation-Parameters",
    "IAU-1980",
    "finals.all",
)

RAD_ARCSEC = pnt.RAD_ARCSEC


@pytest.fixture(autouse=True)
def restore_bundled_eop():
    """Reload the bundled C04 table after each test."""
    yield
    pnt.load_eop_file_data(pnt.get_file_path(C04_FILENAME), True)


def _load_c04():
    pnt.load_eop_file_data(pnt.get_file_path(C04_FILENAME), True)


def _load_finals():
    if not os.path.exists(FINALS_PATH):
        pytest.skip("orekit finals.all snapshot not present")
    pnt.load_eop_finals_file_data(FINALS_PATH, True)


def test_c04_coverage_and_source():
    _load_c04()
    cov = pnt.get_eop_coverage()
    assert cov["source"] == pnt.EopSource.C04
    assert cov["mjd_first"] == pytest.approx(37665.0)  # 1962-01-01
    assert cov["mjd_last"] > cov["mjd_first"]
    # C04 marks non-final rows with a 0.999 formal-error sentinel, so the
    # measured span ends at or before the table end.
    assert cov["mjd_last_measured"] <= cov["mjd_last"]


def test_eop_data_keys_and_units():
    _load_c04()
    cov = pnt.get_eop_coverage()
    data = pnt.get_eop_data(cov["mjd_last_measured"] - 100.0)
    assert set(data) == {
        "x_pole",
        "y_pole",
        "ut1_utc",
        "lod",
        "dpsi",
        "deps",
        "sigma_x_pole",
        "sigma_y_pole",
        "sigma_ut1_utc",
        "dX",
        "dY",
    }
    # C04 carries dpsi/deps, not CIP offsets, so dX/dY are zero for this source.
    assert data["dX"] == 0.0 and data["dY"] == 0.0
    # Polar motion is well under an arcsecond; UT1-UTC is kept under a second by
    # leap seconds; LOD is sub-millisecond.
    assert abs(data["x_pole"] / RAD_ARCSEC) < 1.0
    assert abs(data["y_pole"] / RAD_ARCSEC) < 1.0
    assert abs(data["ut1_utc"]) < 1.0
    assert abs(data["lod"]) < 1e-2
    # The sigmas are formal errors, and on final rows they are far smaller than
    # the values themselves.
    assert 0.0 < data["sigma_x_pole"] / RAD_ARCSEC < 1e-2
    assert 0.0 < data["sigma_ut1_utc"] < 1e-3


def test_out_of_range_epochs_are_clamped_not_extrapolated():
    _load_c04()
    cov = pnt.get_eop_coverage()
    at_end = pnt.get_eop_data(cov["mjd_last"])
    beyond = pnt.get_eop_data(cov["mjd_last"] + 500.0)
    way_beyond = pnt.get_eop_data(cov["mjd_last"] + 5000.0)
    # Held constant: both out-of-range requests return the same endpoint values.
    for key in ("x_pole", "y_pole", "ut1_utc", "lod"):
        assert beyond[key] == pytest.approx(at_end[key], abs=1e-12)
        assert way_beyond[key] == pytest.approx(at_end[key], abs=1e-12)


def test_finals_reader_reaches_further_than_c04():
    _load_c04()
    c04_last = pnt.get_eop_coverage()["mjd_last"]

    _load_finals()
    cov = pnt.get_eop_coverage()
    assert cov["source"] == pnt.EopSource.Finals
    assert cov["mjd_first"] == pytest.approx(41684.0)  # 1973-01-02
    # The point of the finals product: it reaches past the C04 cutoff and keeps
    # going into a predicted span.
    assert cov["mjd_last_measured"] > c04_last
    assert cov["mjd_last"] > cov["mjd_last_measured"]


def test_finals_prediction_flags_are_a_trailing_block():
    _load_finals()
    fd = pnt.get_eop_file_data()
    cov = pnt.get_eop_coverage()
    mjds, is_pred = fd["mjds_utc"], fd["is_prediction"].astype(bool)

    assert fd["source"] == pnt.EopSource.Finals
    assert len(is_pred) == len(mjds)
    assert np.all(np.diff(mjds) > 0)
    assert is_pred.any() and not is_pred.all()
    # Predictions are a single trailing block, delimited by mjd_last_measured.
    assert np.all(np.diff(is_pred.astype(int)) >= 0)
    np.testing.assert_array_equal(is_pred, mjds > cov["mjd_last_measured"])


def test_finals_values_are_physical_including_predictions():
    _load_finals()
    fd = pnt.get_eop_file_data()
    assert np.all(np.abs(fd["x"]) < 1.0)  # ["]
    assert np.all(np.abs(fd["y"]) < 1.0)
    assert np.all(np.abs(fd["ut1_utc"]) < 1.0)  # [s]
    assert np.all(np.abs(fd["lod"]) < 1e-2)


def test_finals_agrees_with_c04_where_both_are_final():
    """MJD 59000 (2020-05-31) is well inside both series' final spans."""
    mjd = 59000.0
    _load_finals()
    finals = pnt.get_eop_data(mjd)
    _load_c04()
    c04 = pnt.get_eop_data(mjd)

    # Tolerances are the genuine C04-vs-Bulletin-A solution difference.
    assert (finals["x_pole"] - c04["x_pole"]) / RAD_ARCSEC == pytest.approx(0.0, abs=5e-3)
    assert (finals["y_pole"] - c04["y_pole"]) / RAD_ARCSEC == pytest.approx(0.0, abs=5e-3)
    assert finals["ut1_utc"] - c04["ut1_utc"] == pytest.approx(0.0, abs=1e-4)
    assert finals["lod"] - c04["lod"] == pytest.approx(0.0, abs=1e-4)


def test_set_eop_source_defaults_to_c04():
    _load_c04()
    assert pnt.get_eop_source() == pnt.EopSource.C04
    assert pnt.get_eop_coverage()["source"] == pnt.EopSource.C04


def test_set_eop_source_switches_an_already_loaded_table():
    """Order independence: selecting a source with a table already loaded must
    switch immediately, not defer (which would amount to ignoring the call)."""
    _load_c04()
    if not os.path.exists(FINALS_PATH):
        pytest.skip("orekit finals.all snapshot not present")

    pnt.set_eop_source(pnt.EopSource.Finals, FINALS_PATH)
    assert pnt.get_eop_source() == pnt.EopSource.Finals
    assert pnt.get_eop_coverage()["source"] == pnt.EopSource.Finals

    pnt.set_eop_source(pnt.EopSource.C04)
    assert pnt.get_eop_coverage()["source"] == pnt.EopSource.C04


def test_set_eop_source_rejects_a_missing_file_and_keeps_the_previous_choice():
    _load_c04()
    with pytest.raises(RuntimeError):
        pnt.set_eop_source(pnt.EopSource.Finals, "/nonexistent/finals.all")
    assert pnt.get_eop_source() == pnt.EopSource.C04
    assert pnt.get_eop_coverage()["source"] == pnt.EopSource.C04


def test_non_forcing_load_is_dropped_when_a_table_is_present():
    """The trap set_eop_source routes around: load_*(force=False) after anything
    has touched EOP is a no-op, leaving the previous product in place."""
    _load_c04()
    if not os.path.exists(FINALS_PATH):
        pytest.skip("orekit finals.all snapshot not present")

    pnt.load_eop_finals_file_data(FINALS_PATH)  # force defaults to False
    assert pnt.get_eop_coverage()["source"] == pnt.EopSource.C04

    pnt.load_eop_finals_file_data(FINALS_PATH, True)
    assert pnt.get_eop_coverage()["source"] == pnt.EopSource.Finals


def test_finals_predicted_sigmas_grow_with_lead_time():
    """Bulletin A publishes its own predicted formal errors, which grow with
    lead time -- the first-cut EOP prediction error budget."""
    _load_finals()
    cov = pnt.get_eop_coverage()
    horizon = cov["mjd_last"] - cov["mjd_last_measured"]
    leads = [lead for lead in (1.0, 30.0, 90.0, 180.0) if lead < horizon]
    sigmas = [pnt.get_eop_data(cov["mjd_last_measured"] + lead)["sigma_ut1_utc"] for lead in leads]

    assert np.all(np.diff(sigmas) > 0), f"sigma_ut1_utc not increasing: {sigmas}"
    # Sanity anchors on the IERS-published magnitudes: sub-ms at a day, several
    # ms at a month.
    assert sigmas[0] < 1e-3
    assert 1e-3 < sigmas[leads.index(30.0)] < 2e-2


# ---------------------------------------------------------------------------
# Step 1 / Step 2a: celestial-pole offsets and the perturbation injector.
# ---------------------------------------------------------------------------

FINALS2000A_PATH = os.path.join(
    pnt.get_data_path(),
    "..",
    "orekit-data-main",
    "Earth-Orientation-Parameters",
    "IAU-2000",
    "finals2000A.all",
)


def test_iau2000_source_populates_cip_offsets():
    """finals2000A carries dX/dY where finals.all carries dpsi/deps; the reader
    must file them under whichever the caller declared."""
    if not os.path.exists(FINALS2000A_PATH):
        pytest.skip("orekit finals2000A.all snapshot not present")
    pnt.load_eop_finals_file_data(FINALS2000A_PATH, True, pnt.EopNutation.Iau2000)
    fd = pnt.get_eop_file_data()
    assert fd["nutation"] == pnt.EopNutation.Iau2000
    assert np.any(fd["dX"] != 0.0)
    assert np.all(fd["dpsi"] == 0.0)  # the unused pair is zero-filled, not left stale
    # CIP offsets are ~0.3 mas rms; anything much larger means dpsi/deps got mis-filed.
    assert np.percentile(np.abs(fd["dX"]), 99) < 5.0  # ["] -- generous, catches a 1000x mix-up
    assert pnt.eop_has_celestial_pole_offsets()

    data = pnt.get_eop_data(59000.0)
    assert abs(data["dX"] / RAD_ARCSEC * 1e3) < 5.0  # mas
    assert abs(data["dY"] / RAD_ARCSEC * 1e3) < 5.0


def test_perturbation_is_additive_and_clearable():
    _load_c04()
    pnt.clear_eop_perturbation()
    assert not pnt.eop_has_celestial_pole_offsets()
    nominal = pnt.get_eop_data(59000.0)

    p = pnt.EopPerturbation()
    p.dx_pole, p.dy_pole, p.dut1, p.dlod = 1e-8, -2e-8, 3e-3, 4e-6
    pnt.set_eop_perturbation(p)
    got = pnt.get_eop_data(59000.0)
    assert got["x_pole"] - nominal["x_pole"] == pytest.approx(1e-8, abs=1e-14)
    assert got["y_pole"] - nominal["y_pole"] == pytest.approx(-2e-8, abs=1e-14)
    assert got["ut1_utc"] - nominal["ut1_utc"] == pytest.approx(3e-3, abs=1e-9)
    assert got["lod"] - nominal["lod"] == pytest.approx(4e-6, abs=1e-12)

    pnt.clear_eop_perturbation()
    restored = pnt.get_eop_data(59000.0)
    assert restored["x_pole"] == pytest.approx(nominal["x_pole"], abs=1e-15)
    assert restored["ut1_utc"] == pytest.approx(nominal["ut1_utc"], abs=1e-12)


def test_perturbation_reaches_the_frame_rotation_with_the_expected_sign():
    """The small-angle convention, pinned by central difference rather than asserted:
    delta r_gcrf = R_itrf->gcrf @ (eps x r_itrf), eps = (-dy_pole, -dx_pole, OMEGA*dut1)."""
    _load_c04()
    omega = 2 * np.pi * 1.00273781191135448 / 86400.0
    t_tdb = pnt.convert_time(pnt.mjd_to_time(59000.0), pnt.Time.UTC, pnt.Time.TDB)
    r_itrf = np.array([4197160.8, 815845.4, 4716876.3])

    def rot(**kw):
        pnt.clear_eop_perturbation()
        if kw:
            p = pnt.EopPerturbation()
            for k, v in kw.items():
                setattr(p, k, v)
            pnt.set_eop_perturbation(p)
        R, _ = pnt.get_frame_rotation_translation(t_tdb, pnt.Frame.ITRF, pnt.Frame.GCRF)
        pnt.clear_eop_perturbation()
        return np.asarray(R, float)

    R = rot()
    # Steps must clear double-precision noise: epochs are ~1.9e9 s since J2000, so a
    # 1e-6 s UT1 step is pure roundoff. See step2a_sensitivity.py.
    for name, step, axis in [
        ("dx_pole", 1e-8, np.array([0.0, -1.0, 0.0])),
        ("dy_pole", 1e-8, np.array([-1.0, 0.0, 0.0])),
        ("dut1", 1e-2, np.array([0.0, 0.0, omega])),
    ]:
        num = (rot(**{name: step}) @ r_itrf - rot(**{name: -step}) @ r_itrf) / (2 * step)
        ana = R @ np.cross(axis, r_itrf)
        rel = np.linalg.norm(num - ana) / np.linalg.norm(ana)
        assert rel < 1e-4, f"{name}: relative error {rel:.2e}"

    pnt.clear_eop_perturbation()
